src/interface_types/resource_handles.rs

```diff
@@ -0,0 +1,371 @@
+        }
+    }
+}
+//////////////////////////////////////////////////////////////////////////////////
+/// HierarchyAuth
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HierarchyAuth {
+    Owner,
+    Platform,
+    Endorsement,
+    Lockout,
+}
+
+impl HierarchyAuth {
+    /// Get the ESYS resource handle for the hierarchy auth.
+    pub fn esys_rh(self) -> ESYS_TR {
+        match self {
+            HierarchyAuth::Owner => ESYS_TR_RH_OWNER,
+            HierarchyAuth::Platform => ESYS_TR_RH_PLATFORM,
+            HierarchyAuth::Endorsement => ESYS_TR_RH_ENDORSEMENT,
+            HierarchyAuth::Lockout => ESYS_TR_RH_LOCKOUT,
+        }
+    }
+
+    /// Get the TPM resource handle for the hierarchy auth.
+    pub fn rh(self) -> TPM2_RH {
+        match self {
+            HierarchyAuth::Owner => TPM2_RH_OWNER,
+            HierarchyAuth::Platform => TPM2_RH_PLATFORM,
+            HierarchyAuth::Endorsement => TPM2_RH_ENDORSEMENT,
+            HierarchyAuth::Lockout => TPM2_RH_LOCKOUT,
+        }
+    }
+}
+
+impl TryFrom<TPM2_HANDLE> for HierarchyAuth {
+    type Error = Error;
+
+    fn try_from(handle: TPM2_HANDLE) -> Result<Self> {
+        match handle {
+            TPM2_RH_OWNER | ESYS_TR_RH_OWNER => Ok(HierarchyAuth::Owner),
+            TPM2_RH_PLATFORM | ESYS_TR_RH_PLATFORM => Ok(HierarchyAuth::Platform),
+            TPM2_RH_ENDORSEMENT | ESYS_TR_RH_ENDORSEMENT => Ok(HierarchyAuth::Endorsement),
+            TPM2_RH_LOCKOUT | ESYS_TR_RH_LOCKOUT => Ok(HierarchyAuth::Lockout),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        }
+    }
+}
+//////////////////////////////////////////////////////////////////////////////////
+/// TODO: HierarchyPolicy
+//////////////////////////////////////////////////////////////////////////////////
+
+//////////////////////////////////////////////////////////////////////////////////
+/// Platform
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Platform {
+    Platform,
+}
+
+impl From<Platform> for AuthHandle {
+    fn from(_: Platform) -> AuthHandle {
+        AuthHandle::PlatformHandle
+    }
+}
+
+impl TryFrom<AuthHandle> for Platform {
+    type Error = Error;
+
+    fn try_from(auth_handle: AuthHandle) -> Result<Platform> {
+        match auth_handle {
+            AuthHandle::PlatformHandle => Ok(Platform::Platform),
+            _ => Err(Error::local_error(WrapperErrorKind::InconsistentParams)),
+        }
+    }
+}
+//////////////////////////////////////////////////////////////////////////////////
+/// Owner
+//////////////////////////////////////////////////////////////////////////////////
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum Owner {
+    Owner,
+    Null, // TODO: Remove null and use Option instead?
```

Comment:
Regarding the TODO, I think keeping a `Null` variant is better, since it's named like that in the spec. Might be confusing to handle it differently here just because it's called `Null`

Reply:
At this point I agree with you (I am not sure I have made up my mind yet) but the reason I thought about it was because of how we sometimes handle ```ESYS_TR_NONE``` as an Option. So I thought the same thing might apply to these null values as well. But as you say the specification mentions the null value explicitly in this types so it might be confusing to remove them.